packages/wallet-adapter-core/src/new/AdaptedWallet.ts

```diff
@@ -0,0 +1,254 @@
+
+  async disconnect() {
+    // TODO: specify which account. defaults to active account
+    const feature = this.features['aptos:disconnect'];
+    await feature.disconnect();
+    for (const callback of this.onAccountDisconnectedListeners) {
+      callback();
+    }
+  }
+
+  // endregion
+
+  // region Accounts
+
+  readonly onAccountConnectedListeners = new Set<(account: AccountInfo) => void>();
+  readonly onAccountDisconnectedListeners = new Set<() => void>();
+
+  async getConnectedAccounts(): Promise<AccountInfo[]> {
+    // TODO: add explicit `getConnectedAccounts` feature
+    const activeAccount = await this.getActiveAccount();
+    return activeAccount ? [activeAccount] : [];
+  }
+
+  onAccountConnected(callback: (account: AccountInfo) => void) {
+    this.onAccountConnectedListeners.add(callback);
+    return () => this.onAccountConnectedListeners.delete(callback);
+  }
+
+  onAccountDisconnected(callback: () => void) {
+    this.onAccountDisconnectedListeners.add(callback);
+    return () => this.onAccountDisconnectedListeners.delete(callback);
+  }
+
+  async getActiveAccount(): Promise<AccountInfo | undefined> {
+    return this.features['aptos:account'].account()
+      .catch(() => undefined);
+  }
+
+  onActiveAccountChanged(callback: (account?: AccountInfo) => void) {
+    const feature = this.features['aptos:onAccountChange'];
+    void feature.onAccountChange((newAccount) => {
+      callback(newAccount);
+    });
+    return () => {
+      void feature.onAccountChange(nullCallback);
+    }
+  }
+
+  // endregion
+
+  // region Networks
+
+  async getAvailableNetworks(): Promise<Network[]> {
+    // TODO: maybe add explicit `getAvailableNetworks` feature
+    return this.availableNetworks;
+  }
+
+  async getActiveNetwork(): Promise<Network> {
+    const feature = this.features['aptos:network'];
+    try {
+      const networkInfo = await feature.network();
+      return networkInfoToNetwork(networkInfo);
+    } catch {
+      return StandardNetwork.MAINNET;
+    }
+  }
+
+  onActiveNetworkChanged(callback: (network?: Network) => void) {
+    const feature = this.features['aptos:onNetworkChange'];
+
+    void feature.onNetworkChange((networkInfo) => {
+      const network = networkInfo && networkInfoToNetwork(networkInfo);
+      callback(network);
+    });
+    return () => {
+      void feature.onNetworkChange(nullCallback);
+    }
+  }
+
+  // endregion
+
+  // region Signature
+
+  // TODO: improve message signature standard
+  async signMessage(input: AptosSignMessageInput) {
+    const feature = this.features['aptos:signMessage'];
+    return feature.signMessage(input);
+  }
+
+  async signTransaction(input: AptosSignTransactionInputV1_1) {
+    const feature = this.features['aptos:signTransaction']
+
+    if (isFeatureMinorVersion(feature, "1.0")) {
+      const { signerAddress, feePayer } = input;
+      // This will throw an error if it requires an async call
+      const transaction = buildTransaction(input);
```

Comment:
hmm looks like we are now requiring the dapp to provide the account sequence number?

Reply:
`signAndSubmitTransaction@1.0` doesn't support payload inputs as arguments.
If the wallet doesn't implement `signAndSubmitTransaction@1.1` and the dapp is passing a payload input, we try our best to make it work.

Better to have the wallet adapter give you an expressive and actionable error, than having the popup not opening on safari :D

I also thought of allowing wallets to specify whether they're ok with async calls through a flag (it can be a feature).
For example, chrome extensions don't have that limitations. Aptos Connect does.

Reply:
> I also thought of allowing wallets to specify whether they're ok with async calls through a flag.
> For example, chrome extensions don't have that limitations. Aptos Connect does.

That could be interesting - I assume since AC has this limitation, other SDK wallets also have it